File: Test/Amazon1.java
package Interviews;

import java.util.*;
public class Amazon1 {

    public static int minimizeMemory(int[] processes, int m){
        int total_sum = 0;
        for(int process : processes){
            total_sum += process;
        }

        int max_sum = 0;
        for(int k = 0; k < m; k ++){
            max_sum += processes[k];
        }

        int i = 0;
        int j = i + m;
        int temp_sum = max_sum;

        
        while(j < processes.length){
            temp_sum -= processes[i];
            temp_sum += processes[j];
            if(temp_sum > max_sum){
                max_sum = temp_sum;
            }
            //temp_sum = Math.max(temp_sum, max_sum);
            i += 1;
            j += 1;


        }
        return total_sum - max_sum;
    }


    public static void main(String[] args) {
        int[] processes = {10,4,8,13,20};
        int m = 2;

        int[] processes1 = {10,4,8,1};
        int m1 = 2;

        int[] processes2 = {4,6,10,8,2,1};
        int m2 = 3;
        System.out.println(minimizeMemory(processes2, m2));
        
    }
}

File: Test/Amazon1.py
def minimizeMemory(processes, m):
    n = len(processes)
    i = 0
    j = i + m
     
    total_sum = sum(processes) ## O(n) 
    max_sum = sum(processes[i:j]) ##O(n)
    tmp_sum = max_sum
    
    while(j < n):  ## O(n-m)
        tmp_sum -= processes[i]
        tmp_sum += processes[j]
        if(tmp_sum > max_sum):
            max_sum = tmp_sum
        
        i += 1
        j += 1 

    return total_sum - max_sum


##O(n)


## Outputs 22
processes = [10,4,8,13,20]
m = 2


processes1 = [10,4,8,1]
m1 = 2

processes2 = [4,6,10,8,2,1]
m2 = 3

assert minimizeMemory(processes, m) == 22
print(True) 

assert minimizeMemory(processes1, m1) == 9
print(True)

assert minimizeMemory(processes2, m2) == 7
print(True)
